pages/tokens.py

```diff
@@ -1,169 +1,241 @@
+
 
 @callback(
-    Output('token-generate', 'n_clicks'),
-    Output('token-table', 'children'),
-    Input('token-generate', 'n_clicks'),
-    State('token-program', 'value'),
-    State('token-length', 'value'),
-    State('token-count', 'value'),
-    State('token-format', 'value'),
-    State('token-checklist', 'value'),
+    Output('generate-tokens-btn', 'disabled'),
+    Output('generate-tokens-alert', 'children'),
+    Output('generate-tokens-alert', 'color'),
+    Input('token-program', 'value'),
+    Input('token-subgroup', 'value'),
+    Input('token-length', 'value'),
+    Input('token-count', 'value'),
 )
-def generate_tokens(n_clicks, program, token_length, token_count, out_format, checklist):
+def validate_token_inputs(program, subgroup, token_length, token_count):
+    if not program:
+        return True, f'Program must be {STUDY_CONFIG or "program"}', 'danger'
+    elif configured_subgroups and (not subgroup or subgroup not in configured_subgroups):
+        return True,f'Subgroup must be one of {configured_subgroups}', 'danger'
+    elif not token_length or token_length < 6 or token_length > 100:
+        return True, 'Token length must be between 6 and 100', 'danger'
+    elif not token_count or token_count < 1:
+        return True, 'Token count must be at least 1', 'danger'
+    
+    example_token = emcao.generate_opcode(token_prefix, program, subgroup, token_length)
+    info = [
+        dbc.Label(
+            f'{token_count} token(s) will be generated in this format:',
+            style={'margin': '0'}
+        ),
+        html.Hr(style={'margin': '8px'}),
+        dbc.Label(
+            example_token,
+            style={'font-family': 'monospace',
+                   'word-break': 'break-all',
+                   'margin': '0' }
+        ),
+    ]
+    return False, info, 'info'
+
+
+@callback(
+    Output('store-tokens', 'data', allow_duplicate=True),
+    Output('generate-tokens-btn', 'n_clicks'),
+    Input('generate-tokens-btn', 'n_clicks'),
+    Input('token-program', 'value'),
+    Input('token-subgroup', 'value'),
+    Input('token-length', 'value'),
+    Input('token-count', 'value'),
+    State('store-tokens', 'data'),
+    Input('store-uuids', 'data'),
+    prevent_initial_call=True
+)
+def generate_tokens(n_clicks, program, subgroup, token_length, token_count, tokens, uuids):
     if n_clicks is not None and n_clicks > 0:
-        token_prefix = get_token_prefix() + program + ('_test' if 'test-token' in checklist else '')
-        tokens = generateRandomTokensForProgram(token_prefix, token_length, token_count, out_format)
-        insert_many_tokens(tokens)
-        for token in tokens:
-            saveAsQRCode(QRCODE_PATH, token)
-    tokens_table = populate_datatable()
-    return 0, tokens_table
+        new_tokens = [
+            emcao.generate_opcode(token_prefix, program, subgroup, token_length)
+            for _ in range(token_count)
+        ]
+        esdt.insert_many_tokens(new_tokens)
```

Comment:
@shankari This line stores the tokens to the token DB

Reply:
Ah I missed that! Phew! 😌 

But actually, although we keep the display/viz functionality in place, we should "remove the store into DB" functionality until we get approval from cyber to have the dashboard be more of a "backoffice" instead of a data/viz solution. We should comment this out for now.